benchmark/utils/timer.hpp

```diff
@@ -0,0 +1,281 @@
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/ginkgo.hpp>
+
+
+#include <chrono>
+#include <memory>
+
+
+#include <gflags/gflags.h>
+
+
+#ifdef HAS_CUDA
+
+
+#include <cuda.h>
+#include <cuda_runtime.h>
+
+
+#include "cuda/base/device_guard.hpp"
+
+
+#endif  // HAS_CUDA
+
+
+#ifdef HAS_HIP
+
+
+#include <hip/hip_runtime.h>
+
+
+#include "hip/base/device_guard.hip.hpp"
+
+
+#endif  // HAS_CUDA
+
+
+DEFINE_bool(gpu_timer, false,
+            "use gpu timer based on event. It is valid only when "
+            "executor is cuda or hip");
+
+
+class Timer {
+public:
+    void tic()
+    {
+        assert(tic_called_ == false);
+        this->tic_impl();
+        tic_called_ = true;
+    }
+
+    std::size_t toc()
+    {
+        assert(tic_called_ == true);
+        auto ns = this->toc_impl();
+        tic_called_ = false;
+        this->add_record(ns);
+        return ns;
+    }
+
+    std::size_t get_total_time() { return total_duration_ns_; }
+
+    std::size_t get_tictoc_num() { return duration_ns_.size(); }
+
+    double get_average_time()
+    {
+        return static_cast<double>(this->get_total_time()) /
+               this->get_tictoc_num();
+    }
+
+    void clear()
+    {
+        duration_ns_.clear();
+        tic_called_ = false;
+        total_duration_ns_ = 0;
+    }
+
+    Timer() : tic_called_(false), total_duration_ns_(0) {}
+
+protected:
+    void add_record(std::size_t ns)
+    {
+        // add the result;
+        duration_ns_.emplace_back(ns);
+        total_duration_ns_ += ns;
+    }
+
+    virtual void tic_impl() = 0;
+
+    virtual std::size_t toc_impl() = 0;
```

Comment:
`toc_impl()` returning is fine with me, but it should be both documented and return `std::int64_t`.